lib/IDE/Refactoring.cpp

```diff
@@ -5397,6 +5404,149 @@ class AsyncConverter : private SourceEntityWalker {
     return true;
   }
 
+  /// Creates an async alternative function that forwards onto the completion
+  /// handler function through
+  /// withCheckedContinuation/withCheckedThrowingContinuation.
+  bool createAsyncWrapper() {
+    assert(Buffer.empty() && "AsyncConverter can only be used once");
+    auto *FD = cast<FuncDecl>(StartNode.get<Decl *>());
+
+    // First add the new async function declaration.
+    addFuncDecl(FD);
+    OS << tok::l_brace << "\n";
+
+    // Then add the body.
+    OS << tok::kw_return << " ";
+    if (TopHandler.HasError)
+      OS << tok::kw_try << " ";
+
+    OS << "await ";
+
+    // withChecked[Throwing]Continuation { cont in
+    if (TopHandler.HasError) {
+      OS << "withCheckedThrowingContinuation";
+    } else {
+      OS << "withCheckedContinuation";
+    }
+    OS << " " << tok::l_brace << " cont " << tok::kw_in << "\n";
+
+    // asyncFn(args...) {
+    addCallToAsyncMethod(FD, TopHandler);
+    OS << " " << tok::l_brace;
+
+    auto SuccessParams = TopHandler.getSuccessParams();
+
+    // Prepare names for the main trailing closure, as well as reference
+    // strings that may include e.g force unwraps to access the value.
+    SmallVector<SmallString<4>, 2> SuccessParamNames;
+    SmallVector<SmallString<8>, 2> SuccessParamRefs;
+    for (auto idx : indices(SuccessParams)) {
+      auto &Param = SuccessParams[idx];
+      SmallString<8> Name;
+      Name.append("res");
+
+      // If we have multiple success params, number them e.g res1, res2...
+      if (SuccessParams.size() > 1)
+        Name.append(std::to_string(idx + 1));
+
+      SuccessParamNames.emplace_back(Name);
+
+      // If we need to unwrap the success param, insert a force unwrap.
+      if (TopHandler.shouldUnwrap(Param.getParameterType()))
+        Name.append("!");
```

Comment:
I guess there's not really much more we can do here... Konrad had the idea of a `fatalError` if both a `nil`, which would at least allow for a better error message. I'll see if anyone else has a better idea.

Reply:
Maybe I'm overdoing it with that idea -- probably the `!` is fine 🤔

Reply:
Given it's generated code, I think a `fatalError` is a little nicer as we can better describe what went wrong.

Reply:
Decided to go with a `guard` and `fatalError`, how does this look?

Reply:
Seems reasonable, if a bit overkill. At least this way it's obvious what the assumptions are. Could have one guard for all the optional params? Though I suppose it's not obvious *which* is nil in that case 🤷.